File: packages/catalog_core/lib/src/ics.dart
/// iCalendar export of active plans: the non-syncing fallback for
/// platforms without calendar integration. Reminders (#74) are all-day
/// VEVENTs without alarm — the calendar shows that something is due;
/// appointments (#75) are timed, with the VALARM the keeper chose.
library;

/// One event for [writeIcs] — already localized and resolved; the
/// core knows entries, not display names.
class IcsEvent {
  /// Stable identity, so re-importing an updated export replaces
  /// events instead of doubling them.
  final String uid;

  /// The due date, rendered as an all-day event — unless [end] is
  /// given, then [date] is the exact local start of a timed event.
  final DateTime date;

  /// Local end of a timed event; null for all-day.
  final DateTime? end;

  final String summary;
  final String description;

  /// A VALARM this many minutes before the start; null for none.
  final int? alertMinutesBefore;

  const IcsEvent(
      {required this.uid,
      required this.date,
      this.end,
      required this.summary,
      required this.description,
      this.alertMinutesBefore});

  bool get allDay => end == null;
}

/// Serializes [events] as an iCalendar file (RFC 5545).
String writeIcs(List<IcsEvent> events, {required DateTime stamp}) {
  final b = StringBuffer()
    ..write('BEGIN:VCALENDAR\r\n')
    ..write('VERSION:2.0\r\n')
    ..write('PRODID:-//cat(a)log//reminders//EN\r\n');
  final dtstamp = _utcStamp(stamp);
  for (final e in events) {
    b
      ..write('BEGIN:VEVENT\r\n')
      ..write(_fold('UID:${_escape(e.uid)}\r\n'))
      ..write('DTSTAMP:$dtstamp\r\n');
    if (e.end case final end?) {
      // Floating local time: the keeper reads it where the visit is.
      b
        ..write('DTSTART:${_local(e.date)}\r\n')
        ..write('DTEND:${_local(end)}\r\n');
    } else {
      final day = e.date;
      final next = day.add(const Duration(days: 1));
      b
        ..write('DTSTART;VALUE=DATE:${_day(day)}\r\n')
        ..write('DTEND;VALUE=DATE:${_day(next)}\r\n');
    }
    b
      ..write(_fold('SUMMARY:${_escape(e.summary)}\r\n'))
      ..write(_fold('DESCRIPTION:${_escape(e.description)}\r\n'));
    if (e.alertMinutesBefore case final minutes?) {
      b
        ..write('BEGIN:VALARM\r\n')
        ..write('ACTION:DISPLAY\r\n')
        ..write(_fold('DESCRIPTION:${_escape(e.summary)}\r\n'))
        ..write('TRIGGER:-PT${minutes}M\r\n')
        ..write('END:VALARM\r\n');
    }
    b.write('END:VEVENT\r\n');
  }
  b.write('END:VCALENDAR\r\n');
  return b.toString();
}

String _day(DateTime d) => '${d.year.toString().padLeft(4, '0')}'
    '${d.month.toString().padLeft(2, '0')}'
    '${d.day.toString().padLeft(2, '0')}';

String _local(DateTime d) => '${_day(d)}T'
    '${d.hour.toString().padLeft(2, '0')}'
    '${d.minute.toString().padLeft(2, '0')}'
    '${d.second.toString().padLeft(2, '0')}';

String _utcStamp(DateTime d) {
  final u = d.toUtc();
  return '${_day(u)}T'
      '${u.hour.toString().padLeft(2, '0')}'
      '${u.minute.toString().padLeft(2, '0')}'
      '${u.second.toString().padLeft(2, '0')}Z';
}

/// TEXT escaping per RFC 5545 §3.3.11.
String _escape(String s) => s
    .replaceAll('\\', '\\\\')
    .replaceAll(';', '\\;')
    .replaceAll(',', '\\,')
    .replaceAll('\r\n', '\\n')
    .replaceAll('\n', '\\n');

/// Content lines longer than 75 octets are folded (RFC 5545 §3.1),
/// splitting only between code points so no UTF-8 sequence is cut.
String _fold(String line) {
  const max = 73; // room for the trailing CRLF within 75
  final body = line.substring(0, line.length - 2);
  final b = StringBuffer();
  var octets = 0;
  var first = true;
  final piece = StringBuffer();
  for (final rune in body.runes) {
    final len = _utf8Len(rune);
    final budget = first ? max : max - 1; // continuation costs a space
    if (octets + len > budget) {
      if (!first) b.write(' ');
      b
        ..write(piece)
        ..write('\r\n');
      piece.clear();
      octets = 0;
      first = false;
    }
    piece.writeCharCode(rune);
    octets += len;
  }
  if (piece.isNotEmpty || first) {
    if (!first) b.write(' ');
    b
      ..write(piece)
      ..write('\r\n');
  }
  return b.toString();
}

int _utf8Len(int rune) => rune < 0x80
    ? 1
    : rune < 0x800
        ? 2
        : rune < 0x10000
            ? 3
            : 4;

File: desktop/crates/catlog-core/src/ics.rs
//! A calendar file: one VEVENT per plan, all-day for reminders, timed
//! with an alarm for appointments, folded at 75 octets as the standard
//! asks. Byte-for-byte what the phone writes.

use chrono::{DateTime, NaiveDateTime, Utc};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct IcsEvent {
    pub uid: String,
    /// The day of an all-day event, or the local start of a timed one.
    pub start: NaiveDateTime,
    /// The local end of a timed event; none for all-day.
    pub end: Option<NaiveDateTime>,
    pub summary: String,
    pub description: String,
    /// A VALARM this many minutes before the start.
    pub alert_minutes_before: Option<u32>,
}

impl IcsEvent {
    pub fn all_day(&self) -> bool {
        self.end.is_none()
    }
}

/// The calendar file for `events`, stamped `stamp`.
pub fn write_ics(events: &[IcsEvent], stamp: DateTime<Utc>) -> String {
    let mut b = String::new();
    b.push_str("BEGIN:VCALENDAR\r\n");
    b.push_str("VERSION:2.0\r\n");
    b.push_str("PRODID:-//cat(a)log//reminders//EN\r\n");
    let dtstamp = format!("{}Z", stamp.naive_utc().format("%Y%m%dT%H%M%S"));
    for e in events {
        b.push_str("BEGIN:VEVENT\r\n");
        b.push_str(&fold(&format!("UID:{}\r\n", escape(&e.uid))));
        b.push_str(&format!("DTSTAMP:{dtstamp}\r\n"));
        match e.end {
            Some(end) => {
                b.push_str(&format!("DTSTART:{}\r\n", local(e.start)));
                b.push_str(&format!("DTEND:{}\r\n", local(end)));
            }
            None => {
                let day = e.start.date();
                let next = day + chrono::Days::new(1);
                b.push_str(&format!("DTSTART;VALUE=DATE:{}\r\n", day.format("%Y%m%d")));
                b.push_str(&format!("DTEND;VALUE=DATE:{}\r\n", next.format("%Y%m%d")));
            }
        }
        b.push_str(&fold(&format!("SUMMARY:{}\r\n", escape(&e.summary))));
        b.push_str(&fold(&format!(
            "DESCRIPTION:{}\r\n",
            escape(&e.description)
        )));
        if let Some(minutes) = e.alert_minutes_before {
            b.push_str("BEGIN:VALARM\r\n");
            b.push_str("ACTION:DISPLAY\r\n");
            b.push_str(&fold(&format!("DESCRIPTION:{}\r\n", escape(&e.summary))));
            b.push_str(&format!("TRIGGER:-PT{minutes}M\r\n"));
            b.push_str("END:VALARM\r\n");
        }
        b.push_str("END:VEVENT\r\n");
    }
    b.push_str("END:VCALENDAR\r\n");
    b
}

fn local(d: NaiveDateTime) -> String {
    d.format("%Y%m%dT%H%M%S").to_string()
}

fn escape(s: &str) -> String {
    s.replace('\\', "\\\\")
        .replace(';', "\\;")
        .replace(',', "\\,")
        .replace("\r\n", "\\n")
        .replace('\n', "\\n")
}

/// Folds one content line (ending in CRLF) at 75 octets: continuation
/// lines start with a space.
fn fold(line: &str) -> String {
    const MAX: usize = 73;
    let body = line.strip_suffix("\r\n").unwrap_or(line);
    let mut out = String::new();
    let mut piece = String::new();
    let mut octets = 0;
    let mut first = true;
    for c in body.chars() {
        let len = c.len_utf8();
        let budget = if first { MAX } else { MAX - 1 };
        if octets + len > budget {
            if !first {
                out.push(' ');
            }
            out.push_str(&piece);
            out.push_str("\r\n");
            piece.clear();
            octets = 0;
            first = false;
        }
        piece.push(c);
        octets += len;
    }
    if !piece.is_empty() || first {
        if !first {
            out.push(' ');
        }
        out.push_str(&piece);
        out.push_str("\r\n");
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::{NaiveDate, TimeZone};

    fn at(y: i32, m: u32, d: u32, h: u32, min: u32) -> NaiveDateTime {
        NaiveDate::from_ymd_opt(y, m, d)
            .unwrap()
            .and_hms_opt(h, min, 0)
            .unwrap()
    }

    #[test]
    fn an_all_day_event_per_plan_folded_within_75_octets() {
        let ics = write_ics(
            &[IcsEvent {
                uid: "catlog-cat-x-f-vaccine@catlog".into(),
                start: at(2029, 5, 4, 0, 0),
                end: None,
                summary: "Miezi — Impfung".into(),
                description: "Auffrischung, mit Umlauten öäü".into(),
                alert_minutes_before: None,
            }],
            Utc.with_ymd_and_hms(2026, 8, 25, 12, 0, 0).unwrap(),
        );
        assert!(ics.starts_with("BEGIN:VCALENDAR\r\nVERSION:2.0\r\n"));
        assert!(ics.contains("DTSTAMP:20260825T120000Z\r\n"));
        assert!(ics.contains("DTSTART;VALUE=DATE:20290504\r\n"));
        assert!(ics.contains("DTEND;VALUE=DATE:20290505\r\n"));
        assert!(ics.contains("SUMMARY:Miezi — Impfung\r\n"));
        assert!(ics.contains("DESCRIPTION:Auffrischung\\, mit Umlauten öäü\r\n"));
        assert!(ics.contains("UID:catlog-cat-x-f-vaccine@catlog\r\n"));
        assert!(!ics.contains("VALARM"));
        assert!(ics.ends_with("END:VEVENT\r\nEND:VCALENDAR\r\n"));
        for line in ics.split("\r\n") {
            assert!(line.len() <= 75, "{line}");
        }
    }

    #[test]
    fn a_timed_event_carries_its_alarm_and_long_lines_fold() {
        let long = "ä".repeat(80);
        let ics = write_ics(
            &[IcsEvent {
                uid: "catlog-appt-1@catlog".into(),
                start: at(2029, 5, 4, 14, 30),
                end: Some(at(2029, 5, 4, 15, 30)),
                summary: "Neutering — 3 cats".into(),
                description: format!("Hugo; Rudi, Minka\n{long}"),
                alert_minutes_before: Some(60),
            }],
            Utc.with_ymd_and_hms(2026, 8, 25, 12, 0, 0).unwrap(),
        );
        assert!(ics.contains("DTSTART:20290504T143000\r\n"));
        assert!(ics.contains("DTEND:20290504T153000\r\n"));
        assert!(!ics.contains("VALUE=DATE"));
        assert!(ics.contains("BEGIN:VALARM\r\nACTION:DISPLAY\r\nDESCRIPTION:Neutering — 3 cats\r\nTRIGGER:-PT60M\r\nEND:VALARM\r\n"));
        assert!(ics.contains(r"DESCRIPTION:Hugo\; Rudi\, Minka\n"));
        let folded: Vec<&str> = ics.split("\r\n").filter(|l| l.starts_with(' ')).collect();
        assert!(!folded.is_empty(), "the long description folds");
        for line in ics.split("\r\n") {
            assert!(line.len() <= 75, "{line}");
        }
        // Unfolding gives the text back.
        let unfolded = ics.replace("\r\n ", "");
        assert!(unfolded.contains(&long));
    }

    #[test]
    fn an_empty_calendar_is_still_a_calendar() {
        let ics = write_ics(&[], Utc.with_ymd_and_hms(2026, 1, 1, 0, 0, 0).unwrap());
        assert_eq!(
            ics,
            "BEGIN:VCALENDAR\r\nVERSION:2.0\r\nPRODID:-//cat(a)log//reminders//EN\r\nEND:VCALENDAR\r\n"
        );
        assert_eq!(fold("X:\r\n"), "X:\r\n");
    }
}
